### DISCUSSION TITLE
[STM32] Interrupts in cmodules do not preempt the main loop

### DISCUSSION THREAD

**Discussion by Meekdai** [micropython/micropython#12828] [Using MicroPython] [answered]
I wrote a cmodules code with stm32f405rgt6 and compiled it successfully. The function of this cmodules code is to switch the level state of GPIO regularly.

The following is the timer initialization code

```
void tim2_init(void)
{
  TIM_HandleTypeDef TIM2_Handle;
  __HAL_RCC_TIM2_CLK_ENABLE();

  // set up and enable interrupt
  NVIC_SetPriority(TIM2_IRQn, NVIC_EncodePriority(NVIC_PRIORITYGROUP_0, 0, 0));
  HAL_NVIC_EnableIRQ(TIM2_IRQn);

  // PWM clock configuration
  TIM2_Handle.Instance = TIM2;
  TIM2_Handle.Init.Period = 65535;
  TIM2_Handle.Init.Prescaler = 0;
  TIM2_Handle.Init.ClockDivision = TIM_CLOCKDIVISION_DIV1;
  TIM2_Handle.Init.CounterMode = TIM_COUNTERMODE_UP;

  HAL_TIM_PWM_Init(&TIM2_Handle);
}

void tim3_init(void)
{
  TIM_HandleTypeDef TIM3_Handle;
  __HAL_RCC_TIM3_CLK_ENABLE();

  // set up and enable interrupt
  NVIC_SetPriority(TIM3_IRQn, NVIC_EncodePriority(NVIC_PRIORITYGROUP_0, 0, 1));
  HAL_NVIC_EnableIRQ(TIM3_IRQn);

  // PWM clock configuration
  TIM3_Handle.Instance = TIM3;
  TIM3_Handle.Init.Period = 65535;
  TIM3_Handle.Init.Prescaler = 0;
  TIM3_Handle.Init.ClockDivision = TIM_CLOCKDIVISION_DIV1;
  TIM3_Handle.Init.CounterMode = TIM_COUNTERMODE_UP;

  HAL_TIM_PWM_Init(&TIM3_Handle);

}
```

After successfully programming the compiled firmware, I successfully saw the GPIO switching. But when I wrote other code in the micropython main loop, I found that the GPIO switching was different. It felt like the timer interrupt did not preempt the main loop.

Whether the interrupt in micropython cannot be preempted?

**Comment by andrewleech** [comment]
If you're using the timer in PWM mode to directly drive the pin, you can do this directly in micropython without using C for timer config
```
PWM_PIN = Pin("A10", Pin.OUT)
PWM_TIMER = Timer(1, freq=1_000_000)
PWM_TIMER_CH = PWM_TIMER.channel(3, Timer.PWM, pin=PWM_PIN, pulse_width_percent=0)
```
If you are only flipping a pin from the timers you should use the directly managed timer channel because then there's no interrupt actually involved, it's all in hardware.

Or are you firing off a Timer interrupt that runs other code, not just pin on/off ?

There is a heirachy to the interrupts, ones with a lower number can / do pre-empt lower number ones in progress: https://github.com/micropython/micropython/blob/64c79a54239d700ecb8d7f07b13388dc79f28993/ports/stm32/irq.h#L134

So it will depend on what code is running in main whether it's doing something that can preempt timer interrupts - looking at that list the obvious one is probably flash.... or lots of uart / usb traffic ?

Considering you're already coding in C land, you can change the IRQ prio on the timer interrupt enable commands, but keep in mind that you might cause issues in the other interrupts if your timer one runs too long. Also if you're writing to internal flash, that completely blocks CPU while in progress so all interrutps get paused.

**Reply by Meekdai** [reply]
Thanks @andrewleech 

I'm porting GRBL into micropython and it works fine when I don't use `socket.send` function.
Now I found that when I send instructions through W5500, the pulse output will stop for 200us, causing the motor to run unstable.
![image](https://github.com/micropython/micropython/assets/11755104/95c26550-6baa-4496-814f-89da58a07937)

**Reply by andrewleech** [reply]
That might be DMA interrupts being used by the SPI communication to W5500.

If you can use the timer.pwm channel to direct drive the gpio without using an interrupt it'll be perfectly stable. Or changing interrupt priorities might be fine too.

**Reply by Meekdai** [reply]
The priority of `DMA` is 5. I changed the priority of `TIM2` to 3, but it still doesn’t work.
```
#define IRQ_PRI_DMA             NVIC_EncodePriority(NVIC_PRIORITYGROUP_4, 5, 0)
```

```
NVIC_SetPriority(TIM2_IRQn, NVIC_EncodePriority(NVIC_PRIORITYGROUP_0, 0, 0));
```

Is it possible that some code in the W5500 disables all interrupts

**Reply by Meekdai** [reply]
I found in `spi_wait_dma_finished`:
https://github.com/micropython/micropython/blob/64c79a54239d700ecb8d7f07b13388dc79f28993/ports/stm32/spi.c#L551C29-L551C29

Is there a way to not disable interrupts?

**Reply by andrewleech** [reply, marked as answer]
That's curious, I'm not really sure why it's needing to disable interrupts like that.... however I'd expect the wfi / re-enable should really kick in for the timer interrupt as well anyway and work without any particular delay to timer interrupt. 
It'd be worth a test with the irq disable/enable lines commented out to just see what happens.

**Reply by Meekdai** [reply]
After commenting these lines, the motor runs smoothly and W5500 communication is also normal.

**Reply by andrewleech** [reply]
git blame on those lines comes up with https://github.com/micropython/micropython/commit/c0496fd44de562cd11e66acd9e42f796c3dcb5fb

Which is interesting reading... but I guess even if the wfi behaves as written I can see it adding some delay / latency as it runs the commands there to exit wfi and re-enable interrupts before the timer interrupt fires off.

Do you know how much delay you're getting on the interrupt? is it like just a bit of jitter (from say these couple of commands running) or is it heaps late / missing interrupts altogether?

**Reply by Meekdai** [reply]
The timer interrupts takes very little time, it runs code like below

```python
void HandleStepResetIT()
{
  GPIO_Write(STEP_GPIO_Port, (GPIO_ReadOutputData(STEP_GPIO_Port) & ~STEP_MASK) | (step_port_invert_mask & STEP_MASK));
}
```

**Reply by andrewleech** [reply]
I was meaning whether you can tell how much the interrupt is delayed by rather than how long it takes to run, trying to figure out if that spi code is causing timer interrupts to be missed or just a fraction late. Probably no way to know unless you can inspect those step output pins with an oscilloscope to see how "late" the pin changes are.

Out of interest, are you using that interrupt to set the gpio levels of multiple pins at the same time?

**Reply by Meekdai** [reply]
https://github.com/orgs/micropython/discussions/12828?notification_referrer_id=NT_kwDOALNeYLM4MjMxNDAwOTkyOjExNzU1MTA0#discussioncomment-7419434

As I mentioned above, the interruption was delayed by about 200us. It can also be seen from the picture. I don't know if you saw the picture I uploaded. It is a screenshot of the oscilloscope.

I found that a configuration of the W5500 also caused this delay, which also disabled interrupts.

https://github.com/micropython/micropython/blob/06a7bf967c58644a0d341764994bdcc9253b4527/extmod/network_wiznet5k.c#L125C1-L131

> Out of interest, are you using that interrupt to set the gpio levels of multiple pins at the same time?

When I run an axis movement, there is only one GPIO switching. If 3 axes are moved at the same time, there will be 3 GPIO switching levels. These GPIO outputs are used as pulse signals for the stepper motor.

**Reply by andrewleech** [reply]
> As I mentioned above, the interruption was delayed by about 200us. It can also be seen from the picture. I don't know if you saw the picture I uploaded. It is a screenshot of the oscilloscope.

Ah, so each pulse on the cro should be a single timer interrupt? So it's not a delay so much as missing a _lot_ of timer interrupts!

If the interrupt is only toggling one gpio I still think you should be able to provide the pin to the timer channel pwm configuration so it can drive it directly without needed an interrupt.  I wouldn't think you even need C for this, look at "PWM Example:" on https://docs.micropython.org/en/latest/library/pyb.Timer.html

**Reply by Meekdai** [reply]
Maybe you haven’t learned about GRBL or 3D printing. The output pulse of each stepper motor needs to be very precise, and the operation between each motor requires some additional algorithms, interpolation/speed look-ahead, etc. These cannot be solved by micropython’s PWM.

Now through some code comments, I have realized that the PC transmits Gcode instructions to micropython through W5500, and micropython calls GRBL (cmodules) to implement interpolation motion.

Thank you very much for your answer.

**Reply by andrewleech** [reply]
Ok sure, nope I've never used a 3d printer nor GRBL, but have used stm32 dma & pwm to generate extremely precise sine waves etc so am used to doing fancy stuff with PWM :-D however I was only going off the info in this thread to think pwm might work for you!

**Reply by Meekdai** [reply]
![image](https://github.com/micropython/micropython/assets/11755104/9392fedd-285d-4f9e-aabf-72c567794a06)

Can stm32 dma & pwm output pulses like the picture above? For example, PA1 outputs 30 pulses to the X axis, and PA2 outputs 26 pulses to the Y axis. They must start and end at the same time.

**Reply by andrewleech** [reply]
I'm going to say yes! but it depends on the speeds and ratio's needed. If you can organise it like 1Mhz tick rate that on each tick the two pins are set on/off to get the overall correct number of changes, that could be done with dma

**Reply by Meekdai** [reply]
Interesting, I always thought it was not possible. I will give it a try when I have time. Can you provide some helpful examples?

**Reply by andrewleech** [reply]
This was a recent project of mine to generate a sine wave with pwm used like a dac. It was a proof of concept to do it in pure python:

https://github.com/andrewleech/micropython-stm32-pwm-sine-wave

Basically use one timer to trigger dma on exact tick rate, the dma then copies values from lookup table onto output register; my output register was pwm duty register but your target register could be just direct to gpio... although that will set all pins on the register I guess which might not help.

Instead you could use dma to update the frequency of two separate timer pwm outputs to get the desired pulse numbers and after xx number of ticks have the frequency set to zero again

**Comment by andrewleech** [comment]
I tried to reproduce this in pure python
``` python
import pyb

pin = pyb.Pin("B12", pyb.Pin.OUT)
def cb(_):
    pin.on()
    pin.off()

tim = pyb.Timer(4)
tim.init(freq=25_000)

spi = pyb.SPI(1, pyb.SPI.CONTROLLER, baudrate=600000, polarity=1, phase=0)

b = "\xA5"*20000

def do_spi():
    for _ in range(10): spi.write(b)

tim.callback(cb)
```
On stm32F411, wIth python  callback I couldn't get faster than ~35Khz timer before the interrupt was using 100% cpu. But with that code I had a pulse stream that when I ran the spi function I'd get some interruptions of ~80uS
Howver, I then commented out all the interrupt stuff in the spi function so it's like:
``` C
STATIC HAL_StatusTypeDef spi_wait_dma_finished(const spi_t *spi, uint32_t t_start, uint32_t timeout) {
    volatile HAL_SPI_StateTypeDef *state = &spi->spi->State;
    for (;;) {
        if (*state == HAL_SPI_STATE_READY) {
            return HAL_OK;
        }
        if (HAL_GetTick() - t_start >= timeout) {
            return HAL_TIMEOUT;
        }
    }
    return HAL_OK;
}
```
But the same 80uS interruptions occurred, no change at all. So I'm not sure where the issue is; perhaps it's an irq priority issue on my end, but even then I'm not seeing the 200uS interruptions you're getting.

**Reply by Meekdai** [reply]
W5500 also has the following functions, which I also commented

https://github.com/micropython/micropython/blob/06a7bf967c58644a0d341764994bdcc9253b4527/extmod/network_wiznet5k.c#L125C1-L131

```C
#if defined(GRBL)
STATIC void wiz_cris_enter(void) {
}

STATIC void wiz_cris_exit(void) {
}
#else
STATIC void wiz_cris_enter(void) {
    wiznet5k_obj.cris_state = MICROPY_BEGIN_ATOMIC_SECTION();
}

STATIC void wiz_cris_exit(void) {
    MICROPY_END_ATOMIC_SECTION(wiznet5k_obj.cris_state);
}
#endif
```
